### PR TITLE
feat(APIM-454): migrate POST `/site` endpoint

### PR DESCRIPTION
**Pull request by avaitonis**
### Introduction
Endpoint POST `/sites` is required to create new sites in Sharepoint. API code will create list item in Sharepoint. Power app will pick up list item and start Sharepoint site creation.

### Resolution
  - POST request payload example `[{ exporterName: 'Acme' }]`
  - Before site creation we check if site already exists
  - If siteService.getSiteStatusByExporterName doesn't throw SiteNotFoundException then response is returned to client
    - Most likely it is information about status of existing Site.
  - If we get SiteNotFoundException, then new site creation can start.
    - Generates new site id using MDM API.
    - Service does MS graph api POST call, with new SiteId for field URL and ExporterName for Title and other fields.
    - Response is based on sharepoint field Sitestatus. In most cases it's value will be `Provisioning` and http status code will be `202`

### CODE REVIEW

File: src/helpers/ukef-id.type.ts
Comment (AlexBramhill) [inline src/helpers/ukef-id.type.ts]: It's a bit surprising that this is `070` and not `0070`. Is this correct?

File: src/modules/site/dto/create-site-request.dto.ts
Comment (AlexBramhill) [inline src/modules/site/dto/create-site-request.dto.ts]: Do we need/are we using this constructor?

File: src/modules/site/exception/site-id-creation-failed.exception.ts
Comment (AlexBramhill) [inline src/modules/site/exception/site-id-creation-failed.exception.ts]: Could we have the standard tests for this please?

File: src/modules/site/site.controller.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.test.ts]: Feels a bit surprising to rely on this generator here -- it'd be good to abstract out the `suteStatusByExporterNameResponse` and call this new generator in both `getSiteStatusByExporterNameGenerator` and `CreateSiteGenerator`. What do you think?

File: src/modules/site/site.controller.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.test.ts]: ('returns a status code of $expectedStatusCode and the expected response if **the created** site status is "$status"')

maybe? Then we can remove the comment as we've made it clear that this test does belong here

File: src/modules/site/site.controller.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.test.ts]: Is this correct?

Comment (AlexBramhill) [reply src/modules/site/site.controller.test.ts]: Ah I think I get this now -- I feel this test could have a rename.

`calls create site and returns a status code of 202 if the site service request to get site status throws a SiteNotFoundException`

File: src/modules/site/site.controller.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.test.ts]: `throws an error if the site service request to get site status throws an error which is not SiteNotFoundException` maybe?

File: src/modules/site/site.controller.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.test.ts]: I think this can just be `siteId`?

File: src/modules/site/site.controller.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.test.ts]: `expectedResponseError`?

File: src/modules/site/site.controller.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.test.ts]: I'd be tempted to split these two controller tests into two files for ease of management

File: src/modules/site/site.controller.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.ts:43]: nice!

File: src/modules/site/site.controller.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.ts]: Could we use an interceptor for this like we have done for the `get` endpoint?

File: src/modules/site/site.controller.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.ts]: @lmarrai-sw might have an opinion on this -- I'd be tempted move this logic to `siteService` as a `createSiteIfDoesNotExist` and simplify the logic of this controller down. There's a lot of nesting here, which I think that moving to siteService will clear up, as well as changing to interceptors

File: src/app.ts
Comment (lmarrai-sw) [inline src/app.ts]: I know it's not part of your change @avaitonis, but we removed `forbidNonWhitelisted: true,` in TFS

File: test/support/mocks/mdm.service.mock.ts
Comment (lmarrai-sw) [inline test/support/mocks/mdm.service.mock.ts]: Very small typo, but did we mean `Mdm` instead of `MdM` here?

File: src/modules/site/site.controller.ts
Comment (avaitonis) [reply src/modules/site/site.controller.ts]: For client we return 500 `Internal server error` without any details.
@luke you suggested rewrap MdmException `Failed to create numbers in MDM` to SiteIdCreationFailedException `Failed to create a site ID`. Is this required at all? 
It's a bit messy in controller, moving to existing SiteExceptionTransformInterceptor is easy, but testing interceptor is not straight forward.

Comment (AlexBramhill) [reply src/modules/site/site.controller.ts]: I believe the usual way we look to test interceptors would be at the api-test level -- would that make it easier?

File: test/support/mocks/mdm.service.mock.ts
Comment (AlexBramhill) [inline test/support/mocks/mdm.service.mock.ts:1]: I'd be tempted as mdm is using simple http endpoints to use `nock` here ala TFS. This allows us the check the request is being sent as expected, and is setting the headers etc as needed.

The reason we don't do this for Graph service is it uses an external package to make these calls. In that case, we don't care about _what_ calls it is making, rather that we are calling the external package correctly, and we mock the response based on that.

File: src/modules/site/site.controller.ts
Comment (lmarrai-sw) [inline src/modules/site/site.controller.ts:59]: Should the return type be `Promise<CreateSiteResponse>` instead?

File: test/common-tests/request-field-validation-api-tests/string-field-validation-api-tests.ts
Comment (lmarrai-sw) [inline test/common-tests/request-field-validation-api-tests/string-field-validation-api-tests.ts]: I noticed a small mistake in TFS this morning: this should be `makeRequest(prepareModifiedRequest(requestWithoutTheField))` instead

Can you fix it here please @avaitonis and I'll fix it in TFS?

**Review by lmarrai-sw — COMMENTED** [review]
Thanks for this @avaitonis

File: src/helpers/ukef-id.type.ts
Comment (avaitonis) [reply src/helpers/ukef-id.type.ts]: Thank you, changed to `0070`.

File: src/modules/site/dto/create-site-request.dto.ts
Comment (avaitonis) [reply src/modules/site/dto/create-site-request.dto.ts]: Removed constructor

File: src/modules/site/exception/site-id-creation-failed.exception.ts
Comment (avaitonis) [reply src/modules/site/exception/site-id-creation-failed.exception.ts]: We had chat with Luke and decided to remove SiteIdCreationFailedException,
Would site id creation fail, then logs will have MdmException.

File: src/modules/site/site.controller.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.test.ts]: Post endpoint is using get site service, so get site generator is also used quite a lot. It's possible to abstract generators, but I am leaving them separate for this PR.

File: src/modules/site/site.controller.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.test.ts]: Thank you, changed test description.

File: src/modules/site/site.controller.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.test.ts]: Thank you, changed test description.

File: src/modules/site/site.controller.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.test.ts]: Thank you, changed test description.

File: src/modules/site/site.controller.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.test.ts]: Changed to shorter version.

File: src/modules/site/site.controller.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.test.ts]: Thanks, changed

File: src/modules/site/site.controller.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.test.ts]: Splitted.

File: src/modules/site/site.controller.ts
Comment (avaitonis) [reply src/modules/site/site.controller.ts]: Thanks Alex, it was quite messy. 

Interceptor worked fine, but was a bit tricky to test. At api-test both exceptions returned same generic 500 error. It was a way to test it, but not so straight forward for me.

We decided to remove SiteIdCreationFailedException.

File: src/modules/site/site.controller.ts
Comment (avaitonis) [reply src/modules/site/site.controller.ts]: Thanks, logic is moved to `siteService`.

File: test/support/mocks/mdm.service.mock.ts
Comment (avaitonis) [reply test/support/mocks/mdm.service.mock.ts:1]: Sounds like good practice, but mdm call is quite simple, so not sure how important it is.

Created new user story APIM-478 so we can prioritise.

File: src/app.ts
Comment (avaitonis) [reply src/app.ts]: OK, I removed forbidNonWhitelisted. One get site test was testing that unknown query parameters not allowed, this test had to be changed.

We could have forbidNonWhitelisted: true, looks like nice extra security measure.

File: test/support/mocks/mdm.service.mock.ts
Comment (avaitonis) [reply test/support/mocks/mdm.service.mock.ts]: Changed

File: src/modules/site/site.controller.ts
Comment (avaitonis) [reply src/modules/site/site.controller.ts:59]: To set http status we need to change object res. Same object is used for response body. It is same approach as in GET site endpoint.

So this function directly is not returning anything.

File: test/common-tests/request-field-validation-api-tests/string-field-validation-api-tests.ts
Comment (avaitonis) [reply test/common-tests/request-field-validation-api-tests/string-field-validation-api-tests.ts]: Fixed

File: src/modules/site/site.controller.create-site.test.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.create-site.test.ts:51]: Nice use of each by the way :)

File: src/modules/site/site.controller.ts
Comment (AlexBramhill) [inline src/modules/site/site.controller.ts:60]: Great job refactoring this

File: test/site/create-post.api-test.ts
Comment (AlexBramhill) [inline test/site/create-post.api-test.ts]: Could the name of this file follow convention from tfs please? (feel free to resolve this comment once updated)

File: test/site/create-post.api-test.ts
Comment (AlexBramhill) [inline test/site/create-post.api-test.ts]: Not needed for now -- but if this test files gets unmanageably large, I think we could probably pull this out into a test part, as it's used in the `get-site-status` endpoint tests too

File: test/site/get-site-status-by-exporter-name.api-test.ts
Comment (AlexBramhill) [inline test/site/get-site-status-by-exporter-name.api-test.ts]: Any reason we are repurposing this test? I'd expect the old test still to pass?

**Review by AlexBramhill — CHANGES_REQUESTED** [review]
Looking nice! Only a couple of comments. (I'd wait until another approval to merge)

File: test/site/get-site-status-by-exporter-name.api-test.ts
Comment (avaitonis) [reply test/site/get-site-status-by-exporter-name.api-test.ts]: Luke highlighted that ValidationPipe configuration in TFS is different, so I removed forbidNonWhitelisted: true. This change broke this GET site test.

File: test/site/create-post.api-test.ts
Comment (avaitonis) [reply test/site/create-post.api-test.ts]: Changed to post-site.api-test.ts . Old name didn't make sense, I just didn't notice.

File: src/modules/site/site.controller.create-site.test.ts
Comment (avaitonis) [reply src/modules/site/site.controller.create-site.test.ts:51]: All yours :)

File: test/site/get-site-status-by-exporter-name.api-test.ts
Comment (lmarrai-sw) [reply test/site/get-site-status-by-exporter-name.api-test.ts]: I think we should either just delete this test or make it "if you use the _correct_ query parameter _and_ an unexpected one as well then it still succeeds/you get a 200 response" - otherwise it's a bit confusing what exactly this test is for :)

File: src/app.ts
Comment (lmarrai-sw) [reply src/app.ts]: `whitelist: true` is a good security measure here and `forbidNonWhitelisted: false` (or missing) is good for client flexibility and release flexibility (@avaitonis and I just discussed this on a call)

File: src/modules/site/site.controller.test.ts
Comment (lmarrai-sw) [reply src/modules/site/site.controller.test.ts]: I'm happy with this

**Review by lmarrai-sw — DISMISSED** [review]
LGTM, thanks @avaitonis

File: test/site/get-site-status-by-exporter-name.api-test.ts
Comment (avaitonis) [reply test/site/get-site-status-by-exporter-name.api-test.ts]: Changed test to "if you use the correct query parameter and an unexpected one as well, then it still succeeds/you get a 202 response". 202 "Provisioning" is default status in get site generator.

**Review by lmarrai-sw — DISMISSED** [review]
LGTM, thanks @avaitonis

File: test/site/get-site-status-by-exporter-name.api-test.ts
Comment (AlexBramhill) [inline test/site/get-site-status-by-exporter-name.api-test.ts]: Is there any reason why this is using this constructor here? We can probably just reference the main generator by the looks of it?

**Review by AlexBramhill — DISMISSED** [review]
lgtm (one slight question)!

File: test/site/get-site-status-by-exporter-name.api-test.ts
Comment (avaitonis) [reply test/site/get-site-status-by-exporter-name.api-test.ts]: Good point, removed redundant generator call.

**Review by AlexBramhill — APPROVED** [review]
Great work!

**Pull request merged**
